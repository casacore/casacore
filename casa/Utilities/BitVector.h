//# BitVector.h: Bit vectors of any size
//# Copyright (C) 1993,1994,1995,1999,2000,2001
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

#ifndef CASA_BITVECTOR_H
#define CASA_BITVECTOR_H

//# Includes
#include <casacore/casa/aips.h>
#include <casacore/casa/Containers/Block.h>
#include <casacore/casa/Utilities/Assert.h>
#include <casacore/casa/iosfwd.h>

namespace casa { //# NAMESPACE CASA - BEGIN

//# Forward Declarations
class BitVectorHelper;

// The size of a unsigned Integer ( assumes 8-bit char )
const uInt WORDSIZE = sizeof(uInt)*8;

// <summary>
// Bit vectors of any size
// </summary>

// <use visibility=export>

// <reviewed reviewer="Friso Olnon" date="1995/03/13" tests="tBitVector" demos="">

// <etymology>
// A variable utilized as a discrete collection of bits is referred
// to as a bit vector.
// </etymology>

// <synopsis> 
// Bit vectors are an efficent method of keeping <em>True/False</em>
// information on a set of items or conditions. Class BitVector
// provides functions to manipulate individual bits in the vector and
// to perform logical operations on whole bit vectors.
// </synopsis> 

// <example>
// <srcblock>
// // Create a bit vector with 20 bits (and set them all to False). 
// BitVector bv (20, False);
//
// // Change some individual bits:
// // Turn On (make True) bit 19.
// bv.setBit (19);
// // Turn Off (make False) bit 12 (superfluous here).
// bv.clearBit (12);
// // Toggle bit 5 (here: change value from 0 (False) to 1 (True)).
// bv.toggleBit (5)
// // Another way of setting a bit using the index operator.
// bv[0] = True;
// // Assign the value of bit 0 to bit 1  (in three ways).
// bv[1] = bv.getBit(0);
// bv[1] = bv[0];
// bv.putBit (1, bv.getBit(0));
//
// // Show the bit vector size and its value on standard output.
// cout << "Size of bit vector:  "<< b.nbits() <<"\n";
// cout << "Value of bit vector: "<< bv        <<"\n";
//
// // Perform logical operations on bit vectors.
// // Create two more bit vectors.
// BitVector bv2 (40, False);
// BitVector bv3 (40, True);
// // bitwise OR
// bv = bv2 | bv3;
// // bitwise AND
// bv = bv2 & bv3;
// // bitwise XOR
// bv = bv2 ^ bv3;
// // bitwise NOT
// bv = ~bv2;
//
// // Reset all bits to False, and then to True
// bv = False;
// bv.set (True);
// // Change the vector's size to 10 (and copy the old values).
// bv.resize (10);
// // Change back to original size and set all bits to True.
// void bv.resize (size, True, False);
// </srcblock>
// </example> 


class BitVector
{
public:
    // BitVectorHelper is a helper class.
    friend class BitVectorHelper;

    // Create a bit vector of length 0.
    BitVector ();

    // Create a bit vector with <src>length</src> bits
    // and set all bits to to the specified state.
    BitVector (uInt length, Bool state);

    // Copy constructor (copy semantics).
    BitVector (const BitVector& that);

    // Delete the bit vector.
   ~BitVector ();

    // Assignment (copy semantics).
    BitVector& operator= (const BitVector& that);

    // Set all bits to the given state.
    BitVector& operator= (Bool state);

    // Return the number of bits in the bitvector.
    uInt nbits() const;

    // Set a bit at the given position (0-relative).
    // In debug-mode an exception is thrown when the position is invalid.
    void setBit (uInt pos); 

    // Clear a bit at the given position (0-relative).
    // In debug-mode an exception is thrown when the position is invalid.
    void clearBit (uInt pos); 

    // Toggle a bit at the given position (0-relative).
    // It returns the original state.
    // In debug-mode an exception is thrown when the position is invalid.
    Bool toggleBit (uInt pos); 

    // Get a bit at the given position (0-relative).
    // In debug-mode an exception is thrown when the position is invalid.
    Bool getBit (uInt pos) const;

    // Set a bit at the given position (0-relative) to the given state.
    // In debug-mode an exception is thrown when the position is invalid.
    void putBit (uInt pos, Bool state);

    // Index operator to access the specified bit.
    // In debug-mode an exception is thrown when the position is invalid.
    // <group>
    Bool operator[] (uInt pos) const;
    BitVectorHelper operator[] (uInt pos);
    // </group>

    // Logical operations on whole bit vectors.
    // The binary operators <src>&</src> (bitwise
    // AND), <src>|</src> (bitwise OR) and <src>^</src> (bitwise XOR),
    // and the unary operator <src>~</src> (bitwise NOT) are provided.
    // An exception is thrown if the lengths of the vectors differ.
    // <group>
    BitVector operator& (const BitVector& that) const;
    BitVector operator| (const BitVector& that) const;
    BitVector operator^ (const BitVector& that) const;
    BitVector operator~ () const;
    // </group>

    // Logical in-place operations on whole bit vectors.
    // The binary operators <src>&</src> (bitwise
    // AND), <src>|</src> (bitwise OR) and <src>^</src> (bitwise XOR),
    // and the unary operator <src>reverse</src> (bitwise NOT) are provided.
    // An exception is thrown if the lengths of the vectors differ.
    // <group>
    void operator&= (const BitVector& that);
    void operator|= (const BitVector& that);
    void operator^= (const BitVector& that);
    void reverse ();
    // </group>

    // Returns True if all bits are equal.
    // An exception is thrown if the lengths of the vectors differ.
    Bool operator== (const BitVector& that) const;

    // Returns True if a bit differs.
    // An exception is thrown if the lengths of the vectors differ.
    Bool operator!= (const BitVector& that) const;

    // Resize the bit vector to the new length.
    // By default the original bits are copied.
    // The remaining bits (or all bits in case of no copy) are
    // set the the given state.
    void resize (uInt length, Bool state=False, Bool copy=True);

    // Set all bits of the bit vector to the specified state.
    void set (Bool state);

    // Set <src>length</src> bits starting at the start position
    // (0-relative) to the given state.
    // An exception is thrown if start+length exceeds the length
    // of the vector.
    void set (uInt start, uInt length, Bool state);

    // Copy <src>length</src> bits starting at thatStart in the
    // other BitVector to this BitVector starting at thisStart.
    void copy (uInt thisStart, uInt length, const BitVector& that,
	       uInt thatStart);

    // Write a representation of the bit vector (a list of
    // <em>zeros</em> and <em>ones</em> enclosed in square
    // parentheses) to ostream.
    friend ostream& operator<< (ostream&, const BitVector& vector);

private:
    // Number of bits in the BitVector object.
    uInt size_p;

    // Pointer to the actual bit vector, stored as a contiguous
    // sequence of one or more unsigned integers.
    Block<uInt> bits_p;
};



// <summary> Helper class for BitVector </summary>
// <use visibility=local>
// <reviewed reviewer="Friso Olnon" date="1995/03/13" tests="tBitVector" demos="">

// <prerequisite>
// <li> class <linkto class=BitVector>BitVector</linkto>
// </prerequisite>

// <synopsis> 
// Helper class for class <linkto class=BitVector>BitVector</linkto>.
// For all practical purposes a BitVectorHelper object is the individual bit in
// a bit vector. It is the object returned by the index operator of
// BitVector.
// </synopsis> 

class BitVectorHelper
{
friend class BitVector;

public:
    // Copy constructor has to be public.
    BitVectorHelper (const BitVectorHelper& that);

    // Set the bit to the state of the bit in the other BitVector.
    // Thus assignment has not the usual copy semantics, but affects
    // the underlying BitVector bit.
    const BitVectorHelper& operator= (const BitVectorHelper& that) const;

    // Set to a state.
    const BitVectorHelper& operator= (Bool state) const;

    // Defines the conversion from <src>BitVectorHelper</src> to
    // <src>Bool</src>.
    operator Bool() const;

private:
    uInt bitNumber_p;

    // Pointer back to the original vector.
    BitVector* vecPtr_p;

    // The constructor we actually use.
    BitVectorHelper (uInt bitNumber, BitVector* vector);
};



inline void BitVector::setBit (uInt pos)
{
    DebugAssert (pos < size_p, AipsError);
    uInt index = pos/WORDSIZE;
    bits_p[index] |= (1 << (pos - index*WORDSIZE));
}

inline void BitVector::clearBit (uInt pos)
{
    DebugAssert (pos < size_p, AipsError);
    uInt index = pos/WORDSIZE;
    bits_p[index] &= (~ (1 << (pos - index*WORDSIZE)));
}

inline Bool BitVector::operator[] (uInt pos) const
{
    return getBit (pos);
}

inline uInt BitVector::nbits() const
{
  return size_p;
}


inline BitVectorHelper::BitVectorHelper (uInt bitNumber, BitVector* vector)
: bitNumber_p (bitNumber),
  vecPtr_p    (vector)
{}

inline BitVectorHelper BitVector::operator[] (uInt pos)
{
    return BitVectorHelper (pos, this);
}

inline BitVectorHelper::BitVectorHelper (const BitVectorHelper& that)
: bitNumber_p (that.bitNumber_p),
  vecPtr_p    (that.vecPtr_p)
{}

inline const BitVectorHelper& BitVectorHelper::operator= (Bool state) const
{
    vecPtr_p->putBit (bitNumber_p, state);
    return *this;
}

inline BitVectorHelper::operator Bool() const
{
    return vecPtr_p->getBit (bitNumber_p);
}

inline const BitVectorHelper& BitVectorHelper::operator=
                                    (const BitVectorHelper& that) const
{
    vecPtr_p->putBit (bitNumber_p, that.vecPtr_p->getBit (that.bitNumber_p));
    return *this;
}




} //# NAMESPACE CASA - END

#endif

