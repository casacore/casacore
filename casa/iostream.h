//# iostream.h: Interim solution for standard/nonstandard system iostream
//# Copyright (C) 2001,2002
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

#ifndef CASA_IOSTREAM_H
#define CASA_IOSTREAM_H

//# Define the C standard C++ include file. 
//# This is an interim solution to cater for the SGI non-existence of
//# them (e.g. <cstring>)

//# Make sure any special macros are set
#include <casacore/casa/aips.h>
#include <iostream>

namespace casacore { //# NAMESPACE CASACORE - BEGIN

using std::istream;
using std::ostream;
using std::iostream;
using std::cin;
using std::cout;
using std::cerr;
using std::endl;
using std::flush;
using std::ws;

using std::ios;
using std::streampos;
using std::streamoff;
using std::streamsize;
using std::dec;
using std::hex;
using std::oct;

using std::internal;
using std::left;
using std::right;
using std::fixed;
using std::scientific;

using std::boolalpha;
using std::noboolalpha;
using std::showbase;
using std::noshowbase;
using std::showpoint;
using std::noshowpoint;
using std::showpos;
using std::noshowpos;
using std::skipws;
using std::noskipws;
using std::uppercase;
using std::nouppercase;

} //# NAMESPACE CASACORE - END

#endif
