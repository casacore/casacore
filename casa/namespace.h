#ifndef NAMESPACE_H
#define NAMESPACE_H 
using namespace casa;
#endif
