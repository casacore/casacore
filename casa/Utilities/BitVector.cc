//# BitVector.cc: A BitVector class with variable bit vector size
//# Copyright (C) 1993,1994,1995,2001
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

#include <casacore/casa/Utilities/BitVector.h>
#include <casacore/casa/iostream.h>


namespace casa { //# NAMESPACE CASA - BEGIN

BitVector::BitVector()
: size_p (0),
  bits_p   (0)
{}

BitVector::BitVector (uInt length, Bool state)
: size_p (length),
  bits_p   ((length + WORDSIZE - 1) / WORDSIZE, uInt(0))
{
    if (state) {
	set (state);
    }
}

BitVector::BitVector (const BitVector& that)
: size_p (that.size_p),
  bits_p   (that.bits_p)
{}

BitVector::~BitVector()
{}

BitVector& BitVector::operator= (const BitVector& that)
{
    size_p = that.size_p;
    bits_p   = that.bits_p;
    return *this;
}

BitVector& BitVector::operator= (Bool state)
{
    set (state);
    return *this;
}

void BitVector::putBit (uInt pos, Bool state)
{
    if (state) {
	setBit (pos);
    }else{
	clearBit (pos);
    }
}

Bool BitVector::toggleBit (uInt pos)
{
    Bool result = getBit (pos);
    putBit (pos,  (!result));
    return result;
}

Bool BitVector::getBit (uInt pos) const
{
    DebugAssert (pos < size_p, AipsError);
    uInt index = pos/WORDSIZE;
    Bool result = True;
    if ((bits_p[index] & (1 << (pos - index*WORDSIZE)))  ==  0) {
	result = False;
    }
    return result;
}

void BitVector::resize (uInt length, Bool state, Bool copy)
{
    //# Do a true resize.
    uInt oldSize = size_p;
    bits_p.resize ((length + WORDSIZE - 1) / WORDSIZE, True, copy);
    size_p = length;
    if (!copy) {
	set (state);
    }else{
	if (length > oldSize) {
	    set (oldSize, length-oldSize, state);
	}
    }
}

void BitVector::set (Bool state)
{
    uInt value = 0;
    if (state) {
	value = ~value;
    }
    for (uInt i=0; i<bits_p.nelements(); i++) {
	bits_p[i] = value;
    }
}
void BitVector::set (uInt start, uInt length, Bool state)
{
    //# Determine the end bit.
    uInt end = start + length;
    if (end > size_p) {
	throw (AipsError ("BitVector::set past end-of-vector"));
    }
    if (length == 0) {
	return;
    }
    //# Determine the full words that can be set.
    //# When setting till the end of the vector, make endWord last word.
    uInt beginWord = (start + WORDSIZE - 1) / WORDSIZE;
    uInt endWord   = end / WORDSIZE;
    if (end == size_p) {
	endWord = bits_p.nelements();
    }
    uInt i;
    //# When there are no full words, we have to do part of a word only.
    if (beginWord >= endWord) {
	for (i=start; i<end; i++) {
	    putBit (i, state);
	}
    }else{
	//# Do some full words and part of the begin and end word.
	uInt value = 0;
	if (state) {
	    value = ~value;
	}
	for (i=beginWord; i<endWord; i++) {
	    bits_p[i] = value;
	}
	for (i=start; i<beginWord*WORDSIZE; i++) {
	    putBit (i, state);
	}
	for (i=endWord*WORDSIZE; i<end; i++) {
	    putBit (i, state);
	}
    }
}

void BitVector::copy (uInt start, uInt length, const BitVector& that,
		      uInt thatStart)
{
    if (start+length > size_p) {
	throw (AipsError ("BitVector::set past end-of-thisvector"));
    }
    if (thatStart+length > that.size_p) {
	throw (AipsError ("BitVector::set past end-of-thatvector"));
    }
    for (uInt i=0; i<length; i++) {
	putBit (start+i, that.getBit (thatStart+i));
    }
}


BitVector BitVector::operator& (const BitVector& that) const
{
    BitVector result(*this);
    result &= that;
    return result;
}
void BitVector::operator&= (const BitVector& that)
{
    if (size_p != that.size_p) {
	throw (AipsError ("BitVector::operator&= with different lengths"));
    }
    for (uInt i=0; i<bits_p.nelements(); i++) {
	bits_p[i] &= that.bits_p[i];
    }
}
BitVector BitVector::operator| (const BitVector& that) const
{
    BitVector result(*this);
    result |= that;
    return result;
}
void BitVector::operator|= (const BitVector& that)
{
    if (size_p != that.size_p) {
	throw (AipsError ("BitVector::operator|= with different lengths"));
    }
    for (uInt i=0; i<bits_p.nelements(); i++) {
	bits_p[i] |= that.bits_p[i];
    }
}
BitVector BitVector::operator^ (const BitVector& that) const
{
    BitVector result(*this);
    result ^= that;
    return result;
}
void BitVector::operator^= (const BitVector& that)
{
    if (size_p != that.size_p) {
	throw (AipsError ("BitVector::operator^= with different lengths"));
    }
    for (uInt i=0; i<bits_p.nelements(); i++) {
	bits_p[i] ^= that.bits_p[i];
    }
}
BitVector BitVector::operator~ () const
{
    BitVector result(*this);
    result.reverse();
    return result;
}
void BitVector::reverse()
{
    for (uInt i=0; i<bits_p.nelements(); i++) {
	bits_p[i] = ~(bits_p[i]);
    }
}


Bool BitVector::operator== (const BitVector& that) const
{
    uInt endWord = size_p / WORDSIZE;
    uInt i;
    for (i=0; i<endWord; i++) {
	if (bits_p[i] != that.bits_p[i]) {
	    return False;
	}
    }
    for (i=endWord*WORDSIZE; i<size_p; i++) {
	if (getBit (i)  !=  that.getBit (i)) {
	    return False;
	}
    }
    return True;
}

Bool BitVector::operator!= (const BitVector& that) const
{
    if (*this == that) {
	return False;
    }
    return True;
}


ostream& operator<< (ostream& os, const BitVector& vector)
{
    // The operator<< write out the less significant bit first
    // (write out first the bit in position zero).
    os << "[";
    for (uInt i=0; i<vector.nbits(); i++) {
	if (vector.getBit (i)) {
            os << "1";
	}else{
            os << "0";
	}
    }
    os << "]" << endl;
    return os;
}

} //# NAMESPACE CASA - END

