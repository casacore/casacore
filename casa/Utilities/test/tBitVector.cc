//# tBitVector.cc: This program tests the BitVector class
//# Copyright (C) 1994,1995,2000,2001
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This program is free software; you can redistribute it and/or modify it
//# under the terms of the GNU General Public License as published by the Free
//# Software Foundation; either version 2 of the License, or (at your option)
//# any later version.
//#
//# This program is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
//# more details.
//#
//# You should have received a copy of the GNU General Public License along
//# with this program; if not, write to the Free Software Foundation, Inc.,
//# 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

//# Includes
#include <casacore/casa/Utilities/BitVector.h>
#include <casacore/casa/iostream.h>

#include <casacore/casa/namespace.h>
void doIt()
{
    uInt maxbits=35;
    // Create bit vectors
    BitVector b;
    BitVector c(maxbits,True);
    BitVector d;
    uInt i;
    b.resize (maxbits);

    for (i=0; i<(maxbits/2); i++)  {
	b.setBit(i*2);
    }
    b[maxbits-2] = b[maxbits-3];
    cout << "b =    ";
    for (i=0; i<maxbits; i++) {
        if (b[i]) {                      // tests operator bool
            cout << "1";
        }else{
            cout << "0";
	}
    }
    cout << endl;

    for (i=0; i<(maxbits/2); i++) {
	c.clearBit(i*2);
    }
    c[maxbits-1] = False;
    cout << "c =   " << c;

    d = b^c; // Operator XOR with two bit vectors
    cout << "b^c = " << d;
    d.set (31, 2, False);
    cout << "d =   " << d;
    cout << "b|c = " << (b|c);
    d = b&c; // Operator AND with two bit vectors
    cout << "b&c = " << d;
    cout << "~b  = " << (~b);

    cout << "b==b = " << (b==b) << endl;
    cout << "b!=b = " << (b!=b) << endl;
    cout << "b==c = " << (b==c) << endl;
    cout << "b!=c = " << (b!=c) << endl;

    BitVector b1(b);
    cout << "b1 = " << b1;
    b1.resize (70);                          // resize longer with copy
    cout << "Size of b1 is "<< b1.nbits() << endl;
    cout << "b1 = " << b1;
    b1.resize (b.nbits());                   // resize shorter with copy
    cout << "b1 = " << b1;
    b1.resize (70, True);                    // resize longer with copy
    cout << "b1 = " << b1;
    b1 = b;
    cout << "b1 = " << b1;
    b1.resize (70, False, False);            // resize without copy
    cout << "b1 = " << b1;
    b1.resize (35, True, False);             // resize without copy
    cout << "b1 = " << b1;

    b1.resize (70, True, False);             // resize without copy
    b1.copy (30, 23, c, 8);
    cout << "b1 = " << b1;
    
    b1 = b;
    for (i=0; i<b.nbits(); i++) {
	b.toggleBit (i);
    }
    if (~b1 != b) {
	cout << "b and ~b1 should be equal" << endl;
    }

    b1.resize (128);
    b1 = False;
    b1.set (32, 64, True);
    for (i=0; i<32; i++) {
	if (b1[i] || !b1[i+32] || !b1[i+64] || b1[i=96]) {
	    cout << "error in b1.set(True,32,64)" << endl;
	}
    }

    BitVector bv1(4,False);
    bv1[0] = True;
    bv1[1] = True;
    BitVector bv2(4,False);
    bv2[0] = True;
    bv2[2] = True;
    BitVector bv3(bv2);
    bv2.reverse();
    if (bv2 != ~bv3  ||  bv2[0] || !bv2[1] || bv2[2] || !bv2[3]) {
	cout << "reverse is incorrect" << endl;
    }
    bv2 = bv3;
    bv2 &= bv1;
    if (bv2 != (bv1&bv3)  ||  !bv2[0] || bv2[1] || bv2[2] || bv2[3]) {
	cout << "=& is incorrect" << endl;
    }
    bv2 = bv3;
    bv2 |= bv1;
    if (bv2 != (bv1|bv3)  ||  !bv2[0] || !bv2[1] || !bv2[2] || bv2[3]) {
	cout << "=| is incorrect" << endl;
    }
    bv2 = bv3;
    bv2 ^= bv1;
    if (bv2 != (bv1^bv3)  ||  bv2[0] || !bv2[1] || !bv2[2] || bv2[3]) {
	cout << "=^ is incorrect" << endl;
    }
}


int main () {
    try {
	doIt();
    } catch (AipsError x) {
	cout << "Caught an exception: " << x.getMesg() << endl;
	return 1;
    } 
    return 0;                           // exit with success status
}
