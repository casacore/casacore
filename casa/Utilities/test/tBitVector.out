b =    10101010101010101010101010101010110
c =   [01010101010101010101010101010101010]
b^c = [11111111111111111111111111111111100]
d =   [11111111111111111111111111111110000]
b|c = [11111111111111111111111111111111110]
b&c = [00000000000000000000000000000000010]
~b  = [01010101010101010101010101010101001]
b==b = 1
b!=b = 0
b==c = 0
b!=c = 1
b1 = [10101010101010101010101010101010110]
Size of b1 is 70
b1 = [1010101010101010101010101010101011000000000000000000000000000000000000]
b1 = [10101010101010101010101010101010110]
b1 = [1010101010101010101010101010101011011111111111111111111111111111111111]
b1 = [10101010101010101010101010101010110]
b1 = [0000000000000000000000000000000000000000000000000000000000000000000000]
b1 = [11111111111111111111111111111111111]
b1 = [1111111111111111111111111111110101010101010101010101011111111111111111]
